File: live/src/main/scala/org/dbpedia/utils/iterator/SinglePrefetchIterator.scala
package org.dbpedia.test;

import org.apache.log4j.Logger
import org.apache.commons.lang.exception.ExceptionUtils;


/**
 * An abstract base class for iterating over containers of unknown size. This
 * works by prefetching junks of the container: Whenever the iterator reaches
 * the end of a chunk, the method "myPrefetch" is called.
 * 
 * 
 * Note that once the iterator is finished (myPrefetch returned null),
 * myPrefetch will never be called again. This means, that if myPrefetch is
 * called, the iterator hasn't reached its end yet.
 * 
 * 
 * @author raven_arkadon
 * @param <T>
 */
abstract class SinglePrefetchIterator[T >: Null <: AnyRef]
	extends Iterator[T]
{
	private val logger : Logger = Logger.getLogger("SinglePrefetchIterator");
	private var current : T = null;
	private var finished : Boolean = false;

	private var advance : Boolean = true;
	
	protected def prefetch() : T;
		//throws Exception;

	/*
	protected SinglePrefetchIterator()
	{
	}
	*/

	protected def finish() : T = {
		this.finished = true;

		return null;
	}
	
	private def _prefetch() : Unit = {
		try {
			current = prefetch();
		}
		catch {
			case e => {
				current = null;		
				logger.error(ExceptionUtils.getFullStackTrace(e));
			}
		}
	}

	
	override def hasNext() : Boolean = {
		if(advance) {
			_prefetch();
			advance = false;
		}

		return finished == false;
	}

	
	override def next() : T =  {
		if(finished) {
			throw new IndexOutOfBoundsException();
		}
		
		if(advance)
			_prefetch();
		
		advance = true;
		return current;
	}

	/*
	override def remove() : Unit = {
		throw new UnsupportedOperationException("Not supported.");
	}
	*/
}

File: live/src/main/java/org/dbpedia/extraction/live/util/iterators/SinglePrefetchIterator.java
package org.dbpedia.extraction.live.util.iterators;

import org.apache.commons.lang3.exception.ExceptionUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Iterator;


/**
 * An abstract base class for iterating over containers of unknown size. This
 * works by prefetching junks of the container: Whenever the iterator reaches
 * the end of a chunk, the method "myPrefetch" is called.
 * <p/>
 * <p/>
 * Note that once the iterator is finished (myPrefetch returned null),
 * myPrefetch will never be called again. This means, that if myPrefetch is
 * called, the iterator hasn't reached its end yet.
 *
 * @param <T>
 * @author raven_arkadon
 */
public abstract class SinglePrefetchIterator<T>
        implements Iterator<T> {
    private static Logger logger = LoggerFactory.getLogger(SinglePrefetchIterator.class);
    private T current = null;
    private boolean finished = false;

    private boolean advance = true;

    protected abstract T prefetch()
            throws Exception;

    protected SinglePrefetchIterator() {
    }

    protected T finish() {
        this.finished = true;

        return null;
    }

    private void _prefetch() {
        try {
            current = prefetch();
        } catch (Exception e) {
            current = null;
            logger.error(ExceptionUtils.getStackTrace(e));
        }
    }

    @Override
    public boolean hasNext() {
        if (advance) {
            _prefetch();
            advance = false;
        }

        return finished == false;
    }

    @Override
    public T next() {
        if (finished) {
            throw new IndexOutOfBoundsException();
        }

        if (advance)
            _prefetch();

        advance = true;
        return current;
    }

    @Override
    public void remove() {
        throw new UnsupportedOperationException("Not supported.");
    }
}
